lib/app.js

```diff
@@ -34,22 +35,53 @@ module.exports = function (options) {
       req.url = path.join('/', host, req.url);
     }
 
-    if (options.cors) {
-      if (req.method === 'OPTIONS') {
-        if (req.headers['access-control-request-headers'])
-          res.header('Access-Control-Allow-Headers', req.headers['access-control-request-headers']);
-        if (req.headers['access-control-request-method'])
-          res.header('Access-Control-Allow-Methods', req.headers['access-control-request-method']);
+    next();
+  });
+
+  let CORSConfiguration;
+  if (typeof options.cors === 'string') {
+    const corsConfigFile = fs.readFileSync(options.cors);
+    xml2js.parseString(corsConfigFile, { async: false }, (err, parsed) => {
+      if (err) {
+        err.name = 'Invalid CORS XML';
+        throw err;
       }
-      if (req.headers.origin)
-        res.header('Access-Control-Allow-Origin', '*');
-    }
+      ({ CORSConfiguration } = parsed);
+    });
+  } else {
+    ({ CORSConfiguration } = options.cors);
+  }
 
+  app.use((req, res, next) => {
+    // Evaluated according to https://docs.aws.amazon.com/AmazonS3/latest/dev/cors.html
+    if (req.method === 'OPTIONS') {
+      res.header('Access-Control-Allow-Origin', '*');
+      if (req.headers['access-control-request-headers'])
+        res.header('Access-Control-Allow-Headers', req.headers['access-control-request-headers']);
+      if (req.headers['access-control-request-method'])
+        res.header('Access-Control-Allow-Methods', req.headers['access-control-request-method']);
+    } else if (CORSConfiguration && CORSConfiguration.CORSRule) {
+      CORSConfiguration.CORSRule.some((rule) => {
+        // The Express CORS middleware is sync, so this finishes before next() is called
+        cors({
+          origin: rule.AllowedOrigin[0] === '*' ? '*' : rule.AllowedOrigin,
+          methods: rule.AllowedMethod,
+          allowedHeaders: rule.AllowedHeader,
+          exposedHeaders: rule.ExposeHeader,
+          maxAge: (rule.MaxAgeSeconds || [])[0]
+        })(req, res, () => {});
+        return !!res.getHeader('Access-Control-Allow-Origin');
```

Comment:
Read the first sentence of the section at the bottom!
> **How Does Amazon S3 Evaluate the CORS Configuration On a Bucket?**
> When Amazon S3 receives a preflight request from a browser, it evaluates the CORS configuration for the bucket and uses the first CORSRule rule that matches the incoming browser request to enable a cross-origin request. For a rule to match, the following conditions must be met:

The `cors` middleware will only set headers when the conditions outlined there match (origin, exposed headers, etc.), so I think I'm still correct in my understanding?

Reply:
I think the exit condition here should be whether the allowed origin matches `rule.AllowedOrigin === req.get('origin')`, not whether the `Access-Control-Allow-Origin` header was set. As long as the first rule has _any_ `AllowedOrigin` here then the header will be set, hence the logic you've added here will use the rules for the first rule rather than the first rule _matching the origin of the incoming browser request_.

To be clearer about the intent this logic (such that there's no ambiguity as to what's going on here and make it obvious if I'm misinterpreting here), I think it would be better to use a `lodash.find` on the `CORSRule` list to explicitly retrieve the first rule matching the request origin and running `cors` with that rule, rather than putting it in a `.some` call and returning your exit condition (really no different from breaking out of a `for` loop).

Reply:
I kind of misspoke, the exit condition is testing that `Access-Control-Allow-Origin` is a truthy value (the `cors` middleware sets it to `false` if it doesn't match). 

This is only basic support for CORS configurations, I offloaded most of the implementation to the `cors` middleware which is a bit imprecise compared to the real S3 response. If you need me to support everything involved in matching a CORSRule (such as partial wildcard `*.example.com` or `http://*.example.com`) I can see about that. I'd probably just ditch the express middleware entirely at that point.

Reply:
Ah I see. I wouldn't have known that without knowing `cors` internals, I'd appreciate a comment on the expectations for the exit condition since they are opaque side effects of an external library. Honestly I'm OK with this as-is, but of course any extra work to support wildcards would be amazing too. Thank you for clarifying.